main: log file

staviq opened the pull request:
WIP implementation of #2694

Logs can be redirected runtime, to a file, stdout or stderr.

https://github.com/staviq/llama.cpp/blob/71d05b9ae438e7d2c3f05212c3dcaf3a53606fa9/examples/main/main.cpp#L61-L73

![image](https://github.com/ggerganov/llama.cpp/assets/15910535/133615cb-71f9-4085-b635-a464b43bbd6f)

~~A way to disable logs completely is currently missing~~, and possibly i could add loglevels too, to filter by debuglevel.

staviq committed:
initial, base LOG macro

staviq committed:
add *.log to .gitignore

staviq committed:
added basic log file handler

staviq committed:
reverted log auto endline to better mimic printf

staviq committed:
remove atomics and add dynamic log target

JohannesGaessler commented:
Somewhat similar PR: https://github.com/ggerganov/llama.cpp/pull/2657

ggerganov commented:
@staviq Yes, this looks similar to what I had in mind - the idea is to use this to generate a trace of the `main` program so we can more easily debug what has happened during some run. The purpose is not to dump structured results - this is the goal of #2657 

You will probably need some sort of helper utils to dump a vector of tokens (e.g. `embd_inp`).

staviq commented:
> Is there an example how to run this along with `main`? Sorry, I don't understand the image. Is it possible to use `main` as usual and include logging?
> 
> `./main -m ~/WizardMath.gguf --color -c 2048 --keep -1 -n -1 -t 3 -b 7 -i -r "USER:" --in-prefix " " -f ~/storage/shared/PT/WizM.txt`
> 
> Thank you.

The screenshot shows `LOG` macro dynamically writing to "screen" and a logfile.

The idea is, that this PR would provide a `LOG` macro, compatible with printf ( format, argumets, etc ) but automatically log to file, or to any valid file handler like stdout or stderr.

I only added a small snippet at the start of `int main()`, in the `main` example, to demonstrate the ability to use LOG while selecting different targets ( a file, stdout,stderr )

The goal is to later use `LOG` in `main` example, to provide better log trace, as mentioned in #2694

At this point, there is one header file `log.h`, and tiny demo in `main.cpp` for testing purposes.

staviq committed:
log_enable/disable, LOG_TEE, basic usage doc

staviq committed:
update .gitignore

staviq committed:
Merge branch 'master' into betterlogs

staviq committed:
mv include to common, params, help msg

staviq committed:
log tostring helpers, token vectors pretty prints

staviq committed:
main: replaced fprintf/LOG_TEE, some trace logging

Cebtenzzre commented on common/log.h:
```diff
@@ -456,3 +481,86 @@ inline void log_print_usage()
+                detokenized.end());                                 \
+                                                                    \
+            buf                                                     \
+                .append("'")                                        \
```
I think this kind of use case is exactly what std::stringstream was made for.

staviq replied:
I had a really exhausting day at work today, so I just pushed what I had before my two remaining brain cells start reinventing the wheel :)

But suggestion much appreciated.

staviq committed:
Merge branch 'ggerganov:master' into betterlogs

ggerganov commented:
Looks very good. When you think the functionality is in a good state, I can make a pass and add the traces that I am mostly interested to `main.cpp` and merge after this

staviq commented:
> Looks very good. When you think the functionality is in a good state, I can make a pass and add the traces that I am mostly interested to `main.cpp` and merge after this

I think I'm gonna wrap all of the helper functions in macros, so in the future they can be removed compile time with a define flag 

I might add some more tostring helpers, but this will become clear when we start using it in `main`.

Other than that, I think the core concept is roughly done.

staviq committed:
LOG_DISABLE_LOGS compile flag, wrapped f in macros

staviq commented:
I think CI is stuck :)

There seems to be a job running continuously for the past 2 months ?
https://github.com/ggerganov/llama.cpp/actions/runs/5368301508

staviq committed:
fix LOG_TEELN and configchecker

staviq committed:
stub LOG_DUMP_CMDLINE for WIN32 for now

staviq commented:
I wanted to un-draft this, but windows seems to hate me, so a bit of help with MSVC would be appreciated, otherwise I'm gonna need some time to find a pc to install windows and msvc on, because the CI logs are not very helpful.

Cebtenzzre commented on common/log.h:
```diff
@@ -0,0 +1,617 @@
+//  invocation with LOG_FILENAME_GENERATOR( "llama", "log" ) creates a string "llama.<number>.log"
+//  where the number is a runtime id of the current thread.
+
+#define LOG_FILENAME_GENERATOR(log_file_basename, log_file_extension) _log_filename_generator(log_file_basename, log_file_extension)
```
I think macros that purely wrap a function should be lowercase. You can consider the fact that it's actually a macro and not a function to be an implementation detail.

Alternatively, you could avoid the macro and have dummy implementations of the functions themselves when LOG_DISABLE_LOGS is defined.

Cebtenzzre replied:
@staviq What do you think?

Cebtenzzre commented on common/log.h:
```diff
@@ -0,0 +1,617 @@
+            buf.append(" ").append(argv[i]);
+        }
+    }
+    LOGLN("Cmd:%s", buf.c_str())
```
This is how MSVC preprocesses this line:
```cpp
{ if (log_handler() != nullptr) { fprintf(log_handler(), "[%lu]" "[%24s:%5d][%24s] " "Cmd:%s", buf.c_str() "%s" , (std::chrono::duration_cast<std::chrono::duration<std::uint64_t>>(std::chrono::system_clock::now().time_since_epoch())).count() , "llama.cpp\\common\\log.h", 511,  __FUNCTION__  ,"\n"); fflush(log_handler()); } }
```
It doesn't like the `buf.c_str() "%s"` part.

staviq replied:
Args are unfolded in a completely wrong place, it seems like msvc interprets __VA_ARGS__ differently.

I managed to clean up enough space for windows VM, currently downloading visual studio. I don't want to abuse fragile CI more then necessary, especially after it broke for over half an hour today :)

Cebtenzzre replied:
This change seems to fix it:
```diff
-#define LOGLN(...) _LOG(__VA_ARGS__, "\n")
+#define LOGLN(str, ...) _LOG(str, ##__VA_ARGS__, "\n")
```

JackJollimore commented:
As mentioned in #2794: logging the tensor information during `main` is probably in the scope of this PR.

I figure @ggerganov check will include logging `llama_model_loader` from `main`, and I think suppressing tensor info. helps readability.

I'll gladly test this PR - I've been checking up on it!

staviq committed:
fix msvc

staviq commented:
Fun fact, IntelliSense and MSVC compiler, on Windows, in the same VisualStudio environment, installed from the same installer, use different preprocessors and have wildly incompatible opinions on variadics :)

I ended up dumping preprocessed code directly with the cl compiler.

No errors now, and none of the warnings from CI are mine.

I'm un-drafting this, and going to sleep, it's already getting bright outside :)

staviq committed:
cleanup main.cpp:273

staviq committed:
Merge branch 'master' into betterlogs

staviq committed:
fix stray whitespace after master sync

staviq commented:
I'm not pushing any more changes so the core concept can get reviewed and approved,

But in the meantime, I figured out how to fairly easily make logging thread-safe, with little to no overhead.

The idea is quite simple, std::atomic<bool> has compare_exchange, with computational cost close to basic if (bool) and it allows to deterministically tell if another invocation is accessing a function, without using a mutex, and only if two threads actually want to use LOG at the exact same time, I can use a mutex, or spin a lambda std::future to queue current log line and let the code continue without waiting. I've been using this in my own projects and I figured it could be used to make logs thread-safe too.

But that's up to you, whether there would be a use case for this.

Cebtenzzre commented on common/log.h:
```diff
@@ -0,0 +1,658 @@
+                _initialized = false;
+            }
+        }
+    }
```
The above chain of conditionals could be made a lot shorter. At least use a boolean `||` for the last two checks: `log_current_filename == filename || log_current_target != target` (I see no reason to use 'compare' here).
What is our code style w/r/t braces? I'd prefer omitting them in the one-line blocks.

Cebtenzzre commented on common/log.h:
```diff
@@ -0,0 +1,658 @@
+            // Log is disabled
+            return nullptr;
+        }
+        else
```
else-after-return is an antipattern.

Cebtenzzre commented on common/log.h:
```diff
@@ -0,0 +1,658 @@
+            }
+            else
+            {
+                logfile = fopen(filename.c_str(), "a");
```
This happens in both the true and false cases, why not place it after the 'if' statement? (And why are we setting logfile to nullptr above before setting it to something else?)

staviq replied:
> And why are we setting logfile to nullptr above before setting it to something else?

My habits from working with C and embedded MCU stuff :) I removed it as it's not necessary in final code.

Cebtenzzre commented on common/log.h:
```diff
@@ -0,0 +1,658 @@
+
+inline bool log_param_single_parse(std::string param)
+{
+    if (std::string("--log-test").compare(param) == 0)
```
Same here - why not just write `if (param == "--log-test")`? C++ is smarter than you think.

Cebtenzzre commented on common/log.h:
```diff
@@ -0,0 +1,658 @@
+        log_test();
+        return true;
+    }
+    else if (std::string("--log-disable").compare(param) == 0)
```
Another return-after-else. `if` is fine here, `else if` is redundant.

ggerganov committed:
Merge branch 'master' into HEAD

ggerganov committed:
log : fix compile warnings

- do not use C++20 stuff
- use PRIu64 to print uint64_t
- avoid string copies by using const ref
- fix ", ##__VA_ARGS__" warnings
- compare strings with == and !=

ggerganov committed:
log : do not append to existing log + disable file line func by default

ggerganov committed:
Merge branch 'master' into HEAD

ggerganov committed:
log : try to fix Windows build

ggerganov committed:
main : wip logs

ggerganov commented:
I think [b97958a](https://github.com/ggerganov/llama.cpp/pull/2748/commits/b97958a5115838bddc635551f42aee89a0937def) broke the Windows build, but I don't have a machine to debug and fix it

ggerganov committed:
main : add trace log

ggerganov reviewed (approved):
After fixing the Windows build we can merge

@JackJollimore 

You can use this branch to run a test that fails for you and send the generated log for analysis. The generated trace is very useful and I've already discovered a bug where we are often using `--in-prefix ' '` which is generally incorrect for the SPM tokenizer. I haven't fixed this in the PR - will do after we merge this and debug the issues that you reported

Here is a sample log from `example/chat.sh`

[main.0x1e5112080.log](https://github.com/ggerganov/llama.cpp/files/12464437/main.0x1e5112080.log)

staviq commented:
I saw the review and comments, and I'll be back at this in couple of hours after work.

JackJollimore commented:
> I've already discovered a bug where we are often using `--in-prefix ' '` which is generally incorrect for the SPM tokenizer. I haven't fixed this in the PR - will do after we merge this and debug the issues that you reported
> 

Lovely. I think I see the bug already:
[main.526799283456.log](https://github.com/ggerganov/llama.cpp/files/12464776/main.526799283456.log)

This is wizardlm q4_0 gguf, here's the content of `M.txt`:
```
Below is an instruction that describes a task. Write a response that appropriately completes the request.

### Instruction:
Please list 3 movie titles.

### Response:
```

reverse prompt `User:` doesn't make it to the console, and there's large sections logging `''`.

I'm re-downloading Vicuna, and WizardMath: **I'll report results** shortly, but it'll take 30 minutes or so to aquire/convert. (I messed up Vicunas tensors 😅)

staviq committed:
Merge branch 'ggerganov:master' into betterlogs

staviq committed:
review: macro f lowercase, str append to sstream

staviq committed:
review: simplify ifs and str comparisons

staviq commented:
Still fighting MSVC. I fixed compilation errors,

But i found a problem where std::isprint() is failing ?

![image](https://github.com/ggerganov/llama.cpp/assets/15910535/5b9292be-7fe3-49ee-a04f-6b000d9c1a25)

Can it be because `llama_token_to_str` was replaced by `llama_token_to_piece` ?

staviq committed:
fix MSVC, formatting, FMT/VAL placeholders

staviq commented:
I went with quick and safe fix for std::isprint, if there is a need for proper unicode-nes in logs, I think It would be less convoluted if I do it in a separate PR.

MSVC is fixed, things appear to be working properly, and if I didn't miss anything, code should be consistent with review suggestions now.

Cebtenzzre commented:
> I didn't miss anything, code should be consistent with review suggestions now.

Eliminate the redundant 'else' clauses after 'if' statements that return in the true case (which I think is what the NOLINT comments are suppressing), and I'll call it good enough.

staviq committed:
review: if/else cleanup

Cebtenzzre commented:
The function returns in both cases above, so this 'else' is unnecessary:
https://github.com/ggerganov/llama.cpp/blob/5cea86927544645ec255a6d74be9794400a7d5ab/common/log.h#L362-L368
(GitHub won't let me make a review comment for some reason.)

staviq committed:
review: if/else cleanup (2)

staviq commented:
I've looked at the logs from clang tidy review CI, and they are all easily doable, do you want me to fix those too ? Shouldn't take too long.

Cebtenzzre commented:
> I've looked at the logs from clang tidy review CI, and they are all easily doable, do you want me to fix those too ?

Sure, go ahead and fix them. I suppose the functions that start with `_` could be changed to end with e.g. `_impl`.

staviq committed:
replace _ prefix with _impl suffix

staviq commented:
I think I just wasted half an hour to do something clang-tidy recommends only to realize it's a bug in clang-tidy :)

It insists on wrapping lambda captures in round brackets, and i couldn't figure out why it doesn't compile, until it hit me: lambda captures cannot be expressions, they have to be identifiers and wrapping them in brackets makes them invalid identifiers :)

ggerganov reviewed (approved):
Thank you! I think these logs will be very helpful for debugging issues in the future

ggerganov commented:
@staviq On MacOS, after adding the `log.h` I'm getting the following compile warnings:

![image](https://github.com/ggerganov/llama.cpp/assets/1991296/ce03aebb-e52d-4d6c-a04e-0d6c8b838599)

At some point, I removed the `','` to fix the warning on my machine, but it made its way back - I guess it breaks something for someone else. Let's look into resolving this

staviq commented:
> @staviq On MacOS, after adding the `log.h` I'm getting the following compile warnings:
> 
> ![image](https://user-images.githubusercontent.com/1991296/268209099-ce03aebb-e52d-4d6c-a04e-0d6c8b838599.png)
> 
> At some point, I removed the `','` to fix the warning on my machine ([b97958a](https://github.com/ggerganov/llama.cpp/commit/b97958a5115838bddc635551f42aee89a0937def)), but it made its way back - I guess it breaks something for someone else. Let's look into resolving this

I don't have a mac to test it on, I'll see if I can get osx to run in VM. But basically, either removing `##` or using the msvc trick should work.

Cebtenzzre commented:
> At some point, I removed the ',' to fix the warning on my machine (b97958a), but it made its way back

Whoops, I added that macro in #3038, actually. I compile with GCC so I didn't see those warnings. Clang will show those warnings on any platform, macOS is not required.
